### PR TITLE
Randomized testing for Join-Exit

### PR DESCRIPTION
**Pull request by mconcat**
Added randomized testing for balancer pool `JoinPool` and `ExitPool`. The added test generates pool with randomized initial assets, and randomized ratio of assets to join/exit. It then performs the sequential JoinPool ExitPool operations, and check if the number of remaining assets and total number of shares are preserved.

## Brief Changelog

*(for example:)*
 
  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/gamm/pool-models/balancer/amm_test.go [hunk 1]
```diff
@@ -436,3 +437,105 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+		initialTokensDenomIn  int64
+		initialTokensDenomOut int64
+
+		percentRatio int64
+
+		numShares sdk.Int
+	}
+
+	const denomOut = "denomOut"
+	const denomIn = "denomIn"
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:452]:
```suggestion
	const (
	  denomOut = "denomOut"
	  denomIn = "denomIn"
	)
```

File: x/gamm/pool-models/balancer/amm_test.go [hunk 2]
```diff
@@ -436,3 +437,105 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+
+	// generate test case with randomized initial assets and join/exit ratio
+	newCase := func() (tc *testCase) {
+		tc = new(testCase)
+		tc.initialTokensDenomIn = rand.Int63n(1_000_000)
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:457]: Should we instead use a source object, i.e. `rng := rand.NewSource(time.Now().Unix())`

Comment (mconcat) [reply x/gamm/pool-models/balancer/amm_test.go]: Replaced!

File: x/gamm/pool-models/balancer/amm_test.go [hunk 3]
```diff
@@ -436,3 +438,111 @@ func TestCalcJoinPoolShares(t *testing.T) {
 		})
 	}
 }
+
+func TestRandomizedPoolInvariants(t *testing.T) {
```
Comment (mattverse) [inline x/gamm/pool-models/balancer/amm_test.go:442]:
```suggestion
func TestRandomizedJoinPoolExitPoolInvariants(t *testing.T) {
```

**Review by mattverse — APPROVED** [review]

File: x/gamm/pool-models/balancer/amm_test.go [hunk 4]
```diff
@@ -436,3 +438,111 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+	}
+
+	const (
+	  denomOut = "denomOut"
+	  denomIn = "denomIn"
+	)
+
+	now := time.Now().Unix()
+	rng := rand.NewSource(now)
+	fmt.Printf("Using random source of %d\n", now)
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:459]:
```suggestion
	t.Logf("Using random source of %d\n", now)
```

**Review by alexanderbez — APPROVED** [review]

**Review by ValarDragon — APPROVED** [review]

File: x/gamm/pool-models/balancer/amm_test.go [hunk 5]
```diff
@@ -436,3 +438,111 @@ func TestCalcJoinPoolShares(t *testing.T) {
 		})
 	}
 }
+
+func TestRandomizedJoinPoolExitPoolInvariants(t *testing.T) {
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/amm_test.go:442]:
```suggestion
// Tests selecting a random amount of coins to LP, and then that ExitPool(JoinPool(tokens))
// preserves the pools number of LP shares, and returns fewer coins to the acter than they started with.
func TestRandomizedJoinPoolExitPoolInvariants(t *testing.T) {
```

File: x/gamm/pool-models/balancer/amm_test.go [hunk 6]
```diff
@@ -436,3 +438,111 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+	// generate test case with randomized initial assets and join/exit ratio
+	newCase := func() (tc *testCase) {
+		tc = new(testCase)
+		tc.initialTokensDenomIn = rng.Int63() % 1_000_000
+		tc.initialTokensDenomOut = rng.Int63() % 1_000_000
+
+		// 1%~100% of initial assets
+		tc.percentRatio = rng.Int63()%100 + 1
+
+		return tc
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/amm_test.go:470]: We should iterate on this selection, but can do in a second PR

**Pull request merged**